<a href="https://colab.research.google.com/github/S-geethika/Python_ML_Codes/blob/main/ASSIGNMENT_5_TFIDF_Submission.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment

<font face='georgia'>
    
   <h4><strong>What does tf-idf mean?</strong></h4>

   <p>    
Tf-idf stands for <em>term frequency-inverse document frequency</em>, and the tf-idf weight is a weight often used in information retrieval and text mining. This weight is a statistical measure used to evaluate how important a word is to a document in a collection or corpus. The importance increases proportionally to the number of times a word appears in the document but is offset by the frequency of the word in the corpus. Variations of the tf-idf weighting scheme are often used by search engines as a central tool in scoring and ranking a document's relevance given a user query.
</p>
    
   <p>
One of the simplest ranking functions is computed by summing the tf-idf for each query term; many more sophisticated ranking functions are variants of this simple model.
</p>
    
   <p>
Tf-idf can be successfully used for stop-words filtering in various subject fields including text summarization and classification.
</p>
    
</font>

<font face='georgia'>
    <h4><strong>How to Compute:</strong></h4>

Typically, the tf-idf weight is composed by two terms: the first computes the normalized Term Frequency (TF), aka. the number of times a word appears in a document, divided by the total number of words in that document; the second term is the Inverse Document Frequency (IDF), computed as the logarithm of the number of the documents in the corpus divided by the number of documents where the specific term appears.

 <ul>
    <li>
<strong>TF:</strong> Term Frequency, which measures how frequently a term occurs in a document. Since every document is different in length, it is possible that a term would appear much more times in long documents than shorter ones. Thus, the term frequency is often divided by the document length (aka. the total number of terms in the document) as a way of normalization: <br>

$TF(t) = \frac{\text{Number of times term t appears in a document}}{\text{Total number of terms in the document}}.$
</li>
<li>
<strong>IDF:</strong> Inverse Document Frequency, which measures how important a term is. While computing TF, all terms are considered equally important. However it is known that certain terms, such as "is", "of", and "that", may appear a lot of times but have little importance. Thus we need to weigh down the frequent terms while scale up the rare ones, by computing the following: <br>

$IDF(t) = \log_{e}\frac{\text{Total  number of documents}} {\text{Number of documents with term t in it}}.$
for numerical stabiltiy we will be changing this formula little bit
$IDF(t) = \log_{e}\frac{\text{Total  number of documents}} {\text{Number of documents with term t in it}+1}.$
</li>
</ul>

<br>
<h4><strong>Example</strong></h4>
<p>

Consider a document containing 100 words wherein the word cat appears 3 times. The term frequency (i.e., tf) for cat is then (3 / 100) = 0.03. Now, assume we have 10 million documents and the word cat appears in one thousand of these. Then, the inverse document frequency (i.e., idf) is calculated as log(10,000,000 / 1,000) = 4. Thus, the Tf-idf weight is the product of these quantities: 0.03 * 4 = 0.12.
</p>
</font>

## Task-1

<font face='georgia'>
    <h4><strong>1. Build a TFIDF Vectorizer & compare its results with Sklearn:</strong></h4>

<ul>
    <li> As a part of this task you will be implementing TFIDF vectorizer on a collection of text documents.</li>
    <br>
    <li> You should compare the results of your own implementation of TFIDF vectorizer with that of sklearns implemenation TFIDF vectorizer.</li>
    <br>
    <li> Sklearn does few more tweaks in the implementation of its version of TFIDF vectorizer, so to replicate the exact results you would need to add following things to your custom implementation of tfidf vectorizer:
       <ol>
        <li> Sklearn has its vocabulary generated from idf sorted in alphabetical order</li>
        <li> Sklearn formula of idf is different from the standard textbook formula. Here the constant <strong>"1"</strong> is added to the numerator and denominator of the idf as if an extra document was seen containing every term in the collection exactly once, which prevents zero divisions.
            
 $IDF(t) = 1+\log_{e}\frac{1\text{ }+\text{ Total  number of documents in collection}} {1+\text{Number of documents with term t in it}}.$
        </li>
        <li> Sklearn applies L2-normalization on its output matrix.</li>
        <li> The final output of sklearn tfidf vectorizer is a sparse matrix.</li>
    </ol>
    <br>
    <li>Steps to approach this task:
    <ol>
        <li> You would have to write both fit and transform methods for your custom implementation of tfidf vectorizer.</li>
        <li> Print out the alphabetically sorted vocab after you fit your data and check if its the same as that of the feature names from sklearn tfidf vectorizer. </li>
        <li> Print out the idf values from your implementation and check if its the same as that of sklearns tfidf vectorizer idf values. </li>
        <li> Once you get your vocab and idf values to be same as that of sklearns implementation of tfidf vectorizer, proceed to the below steps. </li>
        <li> Make sure the output of your implementation is a sparse matrix. Before generating the final output, you need to normalize your sparse matrix using L2 normalization. You can refer to this link https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html </li>
        <li> After completing the above steps, print the output of your custom implementation and compare it with sklearns implementation of tfidf vectorizer.</li>
        <li> To check the output of a single document in your collection of documents,  you can convert the sparse matrix related only to that document into dense matrix and print it.</li>
        </ol>
    </li>
    <br>
   </ul>

  <p> <font color="#e60000"><strong>Note-1: </strong></font> All the necessary outputs of sklearns tfidf vectorizer have been provided as reference in this notebook, you can compare your outputs as mentioned in the above steps, with these outputs.<br>
   <font color="#e60000"><strong>Note-2: </strong></font> The output of your custom implementation and that of sklearns implementation would match only with the collection of document strings provided to you as reference in this notebook. It would not match for strings that contain capital letters or punctuations, etc, because sklearn version of tfidf vectorizer deals with such strings in a different way. To know further details about how sklearn tfidf vectorizer works with such string, you can always refer to its official documentation.<br>
   <font color="#e60000"><strong>Note-3: </strong></font> During this task, it would be helpful for you to debug the code you write with print statements wherever necessary. But when you are finally submitting the assignment, make sure your code is readable and try not to print things which are not part of this task.
    </p>

### Corpus

In [2]:
## SkLearn# Collection of string documents

corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',
]

### SkLearn Implementation

In [3]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()
vectorizer.fit(corpus)
skl_output = vectorizer.transform(corpus)

In [4]:
# sklearn feature names, they are sorted in alphabetic order by default.

print(vectorizer.get_feature_names())

['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']


In [5]:
# Here we will print the sklearn tfidf vectorizer idf values after applying the fit method
# After using the fit function on the corpus the vocab has 9 words in it, and each has its idf value.

print(vectorizer.idf_)

[1.91629073 1.22314355 1.51082562 1.         1.91629073 1.91629073
 1.         1.91629073 1.        ]


In [6]:
# shape of sklearn tfidf vectorizer output after applying transform method.

skl_output.shape

(4, 9)

In [7]:
# sklearn tfidf values for first line of the above corpus.
# Here the output is a sparse matrix

print(skl_output[0])

  (0, 8)	0.38408524091481483
  (0, 6)	0.38408524091481483
  (0, 3)	0.38408524091481483
  (0, 2)	0.5802858236844359
  (0, 1)	0.46979138557992045


In [8]:
# sklearn tfidf values for first line of the above corpus.
# To understand the output better, here we are converting the sparse output matrix to dense matrix and printing it.
# Notice that this output is normalized using L2 normalization. sklearn does this by default.

print(skl_output[0].toarray())

[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]


### Your custom implementation

In [9]:
# Write your code here.
# Make sure its well documented and readble with appropriate comments.
# Compare your results with the above sklearn tfidf vectorizer
# You are not supposed to use any other library apart from the ones given below

from collections import Counter
from tqdm import tqdm
from scipy.sparse import csr_matrix
import math
import operator
from sklearn.preprocessing import normalize
import numpy

In [10]:
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import os

In [11]:
#reference - taken from assignment_BoW_Reference
def fit(dataset):    
    vocab_idf={}
    Doc_frequency_t={}
    unique_words = set() # at first we will initialize an empty set
    # check if its list type or not
    Total_number_of_documents=len(dataset)
    if isinstance(dataset, (list,)):
      
        for row in dataset: # for each review in the dataset
            for word in row.split(" "): # for each word in the review. #split method converts a string into list of words
                if len(word) < 2:
                    continue
                unique_words.add(word)
        unique_words = sorted(list(unique_words))
        vocab = {j:i for i,j in enumerate(unique_words)}

        for word,v in vocab.items():
          cnt=0
          for idx, doc in enumerate((dataset)): 
            word_index = doc.find(word) # retreving the dimension number of a word
            if word_index!=-1:
              cnt+=1
          Doc_frequency_t[word] =cnt
      
        for word,v in vocab.items():
          vocab_idf[word]=1+math.log((1+Total_number_of_documents)/(Doc_frequency_t[word]+1))

        
        return vocab,vocab_idf
    else:
        print("you need to pass list of sentence")

In [12]:
vocab,vocab_idf = fit(corpus)
print()
print("My code feature names")
print(list(vocab_idf.keys()))


My code feature names
['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']


In [13]:
print("idf values after applying the fit method")
print(list(vocab_idf.values()))

idf values after applying the fit method
[1.916290731874155, 1.2231435513142097, 1.5108256237659907, 1.0, 1.916290731874155, 1.916290731874155, 1.0, 1.916290731874155, 1.0]


In [14]:
def transform(dataset,vocab,vocab_idf):
    rows = []
    columns = []
    values = []
    
    #computing Doc_frequency_t
    
    
    for word,v in vocab.items():
      cnt=0
     # for idx, doc in enumerate(tqdm(dataset)): 
      #      word_index = doc.find(word) # retreving the dimension number of a word
       #     if word_index!=-1:
        #      cnt+=1
      #Doc_frequency_t[word] =cnt
      
    #for word,v in vocab.items():
      #IDF1[word]=round(1+math.log((1+Total_number_of_documents)/(Doc_frequency_t[word]+1)),8)
     # IDF[word]=1+math.log((1+Total_number_of_documents)/(Doc_frequency_t[word]+1))

    if isinstance(dataset, (list,)):
    
     
      for idx, row in enumerate((dataset)): # for each document in the dataset
          # it will return a dict type object where key is the word and values is its frequency, {word:frequency}
          #print("idx is ",idx)
          word_freq = dict(Counter(row.split()))
          terms_in_doc=word_freq.values()
          Total_number_of_terms_in_document=sum(terms_in_doc)
      

          # for every unique word in the document
              

          for word, freq in word_freq.items():  # for each unique word in the review.                
              
              if len(word) < 2:
                  continue
              # we will check if its there in the vocabulary that we build in fit() function
              # dict.get() function will return the values, if the key doesn't exits it will return -1
              col_index = vocab.get(word, -1) # retreving the dimension number of a word
              
              # if the word exists
              if col_index !=-1:
                  # we are storing the index of the document
                  rows.append(idx)
                  # we are storing the dimensions of the word
                  columns.append(col_index)
                  #Number_of_t_in_document=freq
                  # we are storing the frequency of the word
                  tf=freq/Total_number_of_terms_in_document
                  freq1=tf*vocab_idf[word]
                  values.append(freq1)
      return csr_matrix((values, (rows,columns)), shape=(len(dataset),len(vocab)))
    else:
        print("you need to pass list of strings")



In [15]:

#sparse matrix
op=transform(corpus, vocab,vocab_idf)
#.toarray()
op1=normalize(op)
print()
print("My code output ")
print("Sparse matrix for first document")
print(op1[0])
print("Dense matrix for first document")
print(op1[0].toarray()) #dense matrix
#[0])


My code output 
Sparse matrix for first document
  (0, 1)	0.4697913855799205
  (0, 2)	0.580285823684436
  (0, 3)	0.3840852409148149
  (0, 6)	0.3840852409148149
  (0, 8)	0.3840852409148149
Dense matrix for first document
[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]


## Task-2

<font face='georgia'>
    <h4><strong>2. Implement max features functionality:</strong></h4>

<ul>
    <li> As a part of this task you have to modify your fit and transform functions so that your vocab will contain only 50 terms with top idf scores.</li>
    <br>
    <li>This task is similar to your previous task, just that here your vocabulary is limited to only top 50 features names based on their idf values. Basically your output will have exactly 50 columns and the number of rows will depend on the number of documents you have in your corpus.</li>
    <br>
    <li>Here you will be give a pickle file, with file name <strong>cleaned_strings</strong>. You would have to load the corpus from this file and use it as input to your tfidf vectorizer.</li>
    <br>
    <li>Steps to approach this task:
    <ol>
        <li> You would have to write both fit and transform methods for your custom implementation of tfidf vectorizer, just like in the previous task. Additionally, here you have to limit the number of features generated to 50 as described above.</li>
        <li> Now sort your vocab based in descending order of idf values and print out the words in the sorted voacb after you fit your data. Here you should be getting only 50 terms in your vocab. And make sure to print idf values for each term in your vocab. </li>
        <li> Make sure the output of your implementation is a sparse matrix. Before generating the final output, you need to normalize your sparse matrix using L2 normalization. You can refer to this link https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html </li>
        <li> Now check the output of a single document in your collection of documents,  you can convert the sparse matrix related only to that document into dense matrix and print it. And this dense matrix should contain 1 row and 50 columns. </li>
        </ol>
    </li>
    <br>
   </ul>

In [16]:
# Below is the code to load the cleaned_strings pickle file provided
# Here corpus is of list type

import pickle
with open('/content/drive/MyDrive/Colab Notebooks/Assignment_5/cleaned_strings', 'rb') as f:
    corpus1 = pickle.load(f)
    
# printing the length of the corpus loaded
print("Number of documents in corpus = ",len(corpus1))

Number of documents in corpus =  746


In [17]:
#reference - taken from assignment_BoW_Reference
#ref - https://www.w3resource.com/python-exercises/dictionary/python-data-type-dictionary-exercise-1.php
def fit1(dataset,max_features):    
    vocab_idf={}
    vocab_idf1={}
    Doc_frequency_t={}
    vocab1={}
    unique_words = set() # at first we will initialize an empty set
    # check if its list type or not
    Total_number_of_documents=len(dataset)
    if isinstance(dataset, (list,)):
      
        for row in dataset: # for each review in the dataset
            for word in row.split(" "): # for each word in the review. #split method converts a string into list of words
                if len(word) < 2:
                    continue
                unique_words.add(word)
        unique_words = sorted(list(unique_words))
        vocab = {j:i for i,j in enumerate(unique_words)}

        for word,v in vocab.items():
          cnt=0
          for idx, doc in enumerate(dataset): 
            word_index = doc.find(word) # retreving the dimension number of a word
            if word_index!=-1:
              cnt+=1
          Doc_frequency_t[word] =cnt
      
        for word,v in vocab.items():
          vocab_idf1[word]=1+math.log((1+Total_number_of_documents)/(Doc_frequency_t[word]+1))
        
        #sort based on idf
        vocab_idf_sorted=dict( sorted(vocab_idf1.items(), key=operator.itemgetter(1),reverse=True))
        vocab_idf=dict(list(vocab_idf_sorted.items())[:max_features])
        vocab1 = { your_key: vocab[your_key] for your_key in  vocab_idf.keys() }
        print(vocab_idf_sorted)
        return vocab1,vocab_idf

    else:
        print("you need to pass list of sentence")

In [18]:
vocab,vocab_idf = fit1(corpus1,50)
print()
print("My code feature names")
print(list(vocab.keys()))

{'aailiyah': 6.922918004572872, 'abandoned': 6.922918004572872, 'abroad': 6.922918004572872, 'abstruse': 6.922918004572872, 'academy': 6.922918004572872, 'accents': 6.922918004572872, 'accessible': 6.922918004572872, 'acclaimed': 6.922918004572872, 'accolades': 6.922918004572872, 'accurately': 6.922918004572872, 'achille': 6.922918004572872, 'ackerman': 6.922918004572872, 'adams': 6.922918004572872, 'added': 6.922918004572872, 'admins': 6.922918004572872, 'admiration': 6.922918004572872, 'admitted': 6.922918004572872, 'adrift': 6.922918004572872, 'adventure': 6.922918004572872, 'aesthetically': 6.922918004572872, 'affected': 6.922918004572872, 'affleck': 6.922918004572872, 'afternoon': 6.922918004572872, 'agreed': 6.922918004572872, 'aimless': 6.922918004572872, 'aired': 6.922918004572872, 'akasha': 6.922918004572872, 'alert': 6.922918004572872, 'alike': 6.922918004572872, 'allison': 6.922918004572872, 'allowing': 6.922918004572872, 'alongside': 6.922918004572872, 'amateurish': 6.92291

In [19]:
print("idf values after applying the fit method")
print(vocab_idf)
#print(list(vocab_idf.values()))

idf values after applying the fit method
{'aailiyah': 6.922918004572872, 'abandoned': 6.922918004572872, 'abroad': 6.922918004572872, 'abstruse': 6.922918004572872, 'academy': 6.922918004572872, 'accents': 6.922918004572872, 'accessible': 6.922918004572872, 'acclaimed': 6.922918004572872, 'accolades': 6.922918004572872, 'accurately': 6.922918004572872, 'achille': 6.922918004572872, 'ackerman': 6.922918004572872, 'adams': 6.922918004572872, 'added': 6.922918004572872, 'admins': 6.922918004572872, 'admiration': 6.922918004572872, 'admitted': 6.922918004572872, 'adrift': 6.922918004572872, 'adventure': 6.922918004572872, 'aesthetically': 6.922918004572872, 'affected': 6.922918004572872, 'affleck': 6.922918004572872, 'afternoon': 6.922918004572872, 'agreed': 6.922918004572872, 'aimless': 6.922918004572872, 'aired': 6.922918004572872, 'akasha': 6.922918004572872, 'alert': 6.922918004572872, 'alike': 6.922918004572872, 'allison': 6.922918004572872, 'allowing': 6.922918004572872, 'alongside':

**Observation:**
All the top 50 feature's IDF value is 6.922918004572872 for both Sklearn and Custom implementation



In [29]:
def transform1(dataset,vocab,vocab_idf):
    rows = []
    columns = []
    values = []
      
    
    if isinstance(dataset, (list,)):
         
     
      for idx, row in enumerate(dataset): # for each document in the dataset
          # it will return a dict type object where key is the word and values is its frequency, {word:frequency}
          #print("idx is ",idx)
          word_freq = dict(Counter(row.split()))
          terms_in_doc=word_freq.values()
          Total_number_of_terms_in_document=sum(terms_in_doc)
         # print(idx,row)
      

          # for every unique word in the document
              

          for word, freq in word_freq.items():  # for each unique word in the review.                
              
              if len(word) < 2:
                  continue
              # we will check if its there in the vocabulary that we build in fit() function
              # dict.get() function will return the values, if the key doesn't exits it will return -1
              
              col_index = vocab.get(word, -1) # retreving the dimension number of a word
              
              # if the word exists
              if col_index !=-1:
                  # we are storing the index of the document
                  rows.append(idx)
                  # we are storing the dimensions of the word
                  columns.append(col_index)
                  #Number_of_t_in_document=freq
                  # we are storing the frequency of the word
                  tf=freq/Total_number_of_terms_in_document
                  freq1=tf*vocab_idf[word]
                  values.append(freq1)

      #print(len(values), len(rows),len(columns),len(dataset),len(vocab))
      return csr_matrix((values, (rows,columns)))
      #, shape=(len(dataset),len(vocab)))
    else:
        print("you need to pass list of strings")

#op=transform1(corpus1, vocab,vocab_idf)
#print(op.shape)

In [30]:

#sparse matrix
op=transform1(corpus1, vocab,vocab_idf)

#.toarray()
op1=normalize(op)
print()
print("My code output ")
print("Sparse matrix for first document")
print(op1[0])
print("Dense matrix for first document")
print(op1[0].toarray()) #dense matrix
#[0])


My code output 
Sparse matrix for first document
  (0, 53)	1.0
Dense matrix for first document
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]


In [23]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer1 = TfidfVectorizer(max_features=50)
vectorizer1.fit(corpus1)
skl_output = vectorizer1.transform(corpus1)

In [31]:
print(vectorizer1.get_feature_names())
s=vectorizer1.idf_

#print(sorted(s,reverse=True))


['acting', 'actors', 'also', 'bad', 'best', 'better', 'cast', 'character', 'characters', 'could', 'even', 'ever', 'every', 'excellent', 'film', 'films', 'funny', 'good', 'great', 'like', 'little', 'look', 'love', 'made', 'make', 'movie', 'movies', 'much', 'never', 'no', 'not', 'one', 'plot', 'real', 'really', 'scenes', 'script', 'see', 'seen', 'show', 'story', 'think', 'time', 'watch', 'watching', 'way', 'well', 'wonderful', 'work', 'would']


In [25]:
skl_output.shape


(746, 50)

In [27]:

print(skl_output[0])

  (0, 25)	1.0


In [28]:
print(skl_output[0].toarray())

[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  0. 0.]]
